client.go

```diff
@@ -0,0 +1,287 @@
+			errcount = 0
+		}
+	}
+}
+
+func (cli *client) Unregister(ctx context.Context, ns string) error {
+	s, err := cli.host.NewStream(ctx, cli.rp, RendezvousProto)
+	if err != nil {
+		return err
+	}
+	defer s.Close()
+
+	w := ggio.NewDelimitedWriter(s)
+	req := newUnregisterMessage(ns, cli.host.ID())
+	return w.WriteMsg(req)
+}
+
+func (cli *client) Discover(ctx context.Context, ns string, limit int, cookie []byte) ([]Registration, []byte, error) {
+	s, err := cli.host.NewStream(ctx, cli.rp, RendezvousProto)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer s.Close()
+
+	r := ggio.NewDelimitedReader(s, inet.MessageSizeMax)
+	w := ggio.NewDelimitedWriter(s)
+
+	return discoverQuery(ns, limit, cookie, r, w)
+}
+
+func discoverQuery(ns string, limit int, cookie []byte, r ggio.Reader, w ggio.Writer) ([]Registration, []byte, error) {
+
+	req := newDiscoverMessage(ns, limit, cookie)
+	err := w.WriteMsg(req)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var res pb.Message
+	err = r.ReadMsg(&res)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if res.GetType() != pb.Message_DISCOVER_RESPONSE {
+		return nil, nil, fmt.Errorf("Unexpected response: %s", res.GetType().String())
+	}
+
+	status := res.GetDiscoverResponse().GetStatus()
+	if status != pb.Message_OK {
+		return nil, nil, RendezvousError{Status: status, Text: res.GetDiscoverResponse().GetStatusText()}
+	}
+
+	regs := res.GetDiscoverResponse().GetRegistrations()
+	result := make([]Registration, 0, len(regs))
+	for _, reg := range regs {
+		pi, err := pbToPeerInfo(reg.GetPeer())
+		if err != nil {
+			log.Errorf("Invalid peer info: %s", err.Error())
+			continue
+		}
+		result = append(result, Registration{Peer: pi, Ns: reg.GetNs(), Ttl: int(reg.GetTtl())})
+	}
+
+	return result, res.GetDiscoverResponse().GetCookie(), nil
+}
+
+func (cli *client) DiscoverAsync(ctx context.Context, ns string) (<-chan Registration, error) {
+	s, err := cli.host.NewStream(ctx, cli.rp, RendezvousProto)
+	if err != nil {
+		return nil, err
+	}
+
+	ch := make(chan Registration)
+	go discoverAsync(ctx, ns, s, ch)
+	return ch, nil
+}
+
+func discoverAsync(ctx context.Context, ns string, s inet.Stream, ch chan Registration) {
+	defer s.Close()
+	defer close(ch)
+
+	r := ggio.NewDelimitedReader(s, inet.MessageSizeMax)
+	w := ggio.NewDelimitedWriter(s)
+
+	const batch = 200
+
+	var (
+		cookie []byte
+		regs   []Registration
+		err    error
+	)
+
+	for {
+		regs, cookie, err = discoverQuery(ns, batch, cookie, r, w)
+		if err != nil {
+			// TODO robust error recovery
+			//      - handle closed streams with backoff + new stream, preserving the cookie
+			//      - handle E_INVALID_COOKIE errors in that case to restart the discovery
+			log.Errorf("Error in discovery [%s]: %s", ns, err.Error())
+			return
+		}
+
+		for _, reg := range regs {
+			select {
+			case ch <- reg:
+			case <-ctx.Done():
+				return
+			}
+		}
+
+		if len(regs) < batch {
+			// TODO adaptive backoff for heavily loaded rendezvous points
```

Comment:
Not sure if a load indicator would help, as the clients can simply ignore it.
I am in favour of load inference from the rate of registration responses.

Reply:
On the other hand, we could make this an Error response -- `E_TEMPORARILY_UNAVAILABLE` perhaps.

Now that could help, as it would force the clients to back off and possibly create a new stream.

Reply:
Yeah, I was thinking have it be an error response. So they don't get any records back, and the error tells them to just wait and try again in a bit